### ISSUE TITLE
Feature: "Smart" Playlists

### ISSUE THREAD

**Issue by calne-ca** [airsonic/airsonic#803]
It would be really cool if airsonic would support smart/dynamic playlists. With "smart" playlists I mean playlists that are not composed out of a fixed amount of songs but rather from a set of rules/filters that automatically fill the playlist with all songs that comply to these rules and filters. For example one could create a playlist that contains all songs from a set of directories:

J-Pop (smart playlist)
- /media/REOL
- /media/AKB48
- /media/YUI

This example playlist would add every song under the sub directories REOL, AKB48 and YUI and add them to the playlist.

Smart playlists have several advantages over normal playlists:
- They automatically (after every scan) add new songs to the playlists if they comply to the playlists' rules/filters 
- They automatically adapt to any filesystem change as long as it occurs under the selected directories. This includes the renaming and removal of sub directories and files.
- If you have songs/artists that you want to add to multiple playlists you do not have to adapt every single playlist every time you add a new song.

In other words: You do not need to trouble yourself with maintaining your playlists.

AIMP implemented smart playlists about 3 years ago and it is really one of its best features. I've used it for a year now and I rarely need to adapt any of my playlists.

I only mentioned the declaration of directories but of course it's also a possibility to use regex for filtering via the songs' meta data or file name.

**Comment by adamdmoss** [comment]
FYI I use beets to do something similar but more flexible - I generate smart playlists from queries and put them into playlists files which airsonic knows how to refresh.  This way I can mark files as belonging to one or more playlists explicitly, and combine these with queries based on filename, newness, samplerate, etc:

from beets/config.yaml:
```
smartplaylist:
    auto: no
    relative_to: 
    playlist_dir: /music/playlists
    playlists:
          # MANUALLY TAGGED:
        - name: g-instrumental.m3u
          query: '_playlist-instrumental:x'
        - name: g-48khz-24bit.m3u
          query: 'samplerate:48000.. bitdepth:24..'
        - name: g-48khz.m3u
          query: 'samplerate:48000..'
        - name: g-mass-effect.m3u
          query: 'mass effect'
          # MANUALLY TAGGED:
        - name: g-inoffensive-dinner.m3u
          query: '_playlist-inoffensive-dinner:x'
```
and so on.

**Comment by calne-ca** [comment]
Thank you very much for this information. I never heard of beets before but I tried it just now and after a little bit of configuration it seems to work. I am now able to create smart playlists and automatically put the resulting .m3u files (which contain the correct paths for airsonic) into airsonic's palylist directory. Airsonic even says "Completed playlist import" in the log but they don't actually appear in the Web UI. I am also kinda wondering how the airsonic playlist import works at all, since playlists are user specific how should airsonic know which user the playlist belongs to? The imported playlist isn't listed in any of the existing users playlists. Do you know what I'm doing wrong?

**Comment by fxthomas** [comment]
I tried having another look at this.

I was thinking that it could be a good idea to explore having some sort of "advanced" search syntax (like [Beet](https://beets.readthedocs.io/en/stable/reference/query.html) or [Lucene](https://lucene.apache.org/core/8_2_0/queryparser/org/apache/lucene/queryparser/classic/QueryParser.html)) that could be used to search for `starred:true genre:Rock artist:"the*"` (for searching for all starred Rock songs whose artist starts with "the").

This in turn would make implementing smart playlists relatively easy (we just store the query and use it when playing from the smart playlist).

The other option (which I like less, because it's much less flexible) is to add a column for each search term in smart playlists.

------

I can see a few choices for this in terms of technology:

* (A) Use the database entirely
  - This would be very similar to how "shuffle radio" currently behaves
  - This has the least chances of having data not up-to-date
  - The performance could be a little lower than other solutions (though not necessarily, we need to measure that)
* (B) Use Lucene entirely
  - This reuses and extends the existing search infrastructure
  - We need to add other fields to the Lucene index
  - We can reuse the Lucene query parser for our search language
  - **Some fields (e.g. "starred" status) are updated outside of scans**
* (C) Use a mix of both
  - More complex, more chances of failure, potentially much less performant.

I very much like option (B), especially given the advantages, but it has some potential blocking points (in bold).

@tesshucom I know you've worked on this a lot (also I just noticed https://github.com/tesshucom/jpsonic/issues/364) : do you know if it is possible to partially update the index (such as when starring a song)? Do you have an idea of the impact on the stability if we implemented such a thing?

**Comment by tesshucom** [comment]
You can read expert documents.
[Subsonic utilise Elasticsearch comme base de données orientée documents](https://www.elastic.co/fr/blog/subsonic-utilise-elasticsearch-comme-base-de-donnees-orientee-documents)
This document describes approach (B).

>  do you know if it is possible to partially update the index (such as when starring a song)?

Partial updates are possible.

What I am personally promoting is the approach to (C).

___

There are two types of server data update: synchronous processing (DB) and asynchronous processing (Lucene).
If you focus on the search function and increase your dependency on Lucene, there are some processes that are more difficult.

It is desirable that partial update should be an API that can be called back.
For example, if you click one star on the screen, after the data is updated on the server, all identical stars on the client's screen will be inverted.

If you really want to do this.

 - Airsonic supports xml and json, while Subsonic also supports jsonp.
In this regard, Subsonic may be pursuing a more realistic design than Airsonic.
 - If Airsonic indiscriminately replaces variables with Lucene, the hurdle for maintaining consistency goes up.

Of course, although it is possible to perform synchronous processing at the plan (B), it is considerably complicated implementation.
(Call back after DB and Lucene are completely updated. Star updates are still very slow today, but will be slower than now.)

___

I think the "potential blocking point (bold)" presented by fxthomas is exactly the point of design.
If priority is placed on ease of verification, the following may be considered.

 - Distributed processing could avoid difficult synchronization problems, Maintenance is easier if strict processing is performed in the DB.
 - Fields to be updated outside of scan should not be included in Lucene implementation.
 - Processes that should not be performed in DB are performed on Lucene side (ex :  #1217).

This is a personal idea.
And I think that what is made according to this policy will be more complicated than it is now.
I don't care much about it.
I want a design that is consistent when tested and does not suffer from asynchronous bugs in the future.

___

Smart playlists don't seem too difficult as long as the grammar is clear.
I think the implementation of UPnP search done in my clone might be a bit similar.

We can define the [grammar](https://github.com/tesshucom/jpsonic/blob/master/jpsonic-upnp-template/src/main/antlr4/com/tesshu/jpsonic/service/upnp/UPnPSearchCriteria.g4) using [antlr](https://www.antlr.org/), analyze it with the [Director](https://github.com/tesshucom/jpsonic/blob/master/jpsonic-main/src/main/java/org/airsonic/player/service/search/UPnPCriteriaDirector.java) class, and convert it to a Lucene query at the same time.

**Comment by fxthomas** [comment]
Thank you, that is definitely useful. I will probably need some time to digest all this and test things though! Will probably wait before 11.0 is released :wink:

**Comment by tesshucom** [comment]
In fact, (A), (B) and (C) each have problems.
I value the direction of (C), but it is not absolute.
I'm looking for a method that is convenient and consistent when using Japanese.

Although my clone has many additions related to meta-processing, I am not trying to make anything above Airsonic.
It is tailored to the needs of the Japanese.

It may be difficult to get started right away.
But if you have some parts of interesting features or code, you can reuse them with Airsonic.
When porting to Airsonic, please change the header as you like and duplicate as you like.
(I, can save the trouble of PR)

The function added in the clone is marked on the setting screen.
It's a little easier to read the settings screen than to look at the code.

**Comment by deusxanime** [comment]
Subscribing and adding +1 to this. I've been using Airsonic for a while but recently tried out Plex for my music and they have smart playlists. I have to say it is a killer feature. In Airsonic when I added a new song or songs, I'd have to update multiple playlists to add that new song into them. In Plex I created smart playlists with filters based on genre, artist, rating, etc and all I have to do now is make sure my metadata is set correctly on anything I add (which I always do anyway, I'm particular about it), scan it in, and it is automatically added to all my playlists as needed with no effort or remembering to do so required from me.

**Comment by undersound** [comment]
I already love airsonic and this feature will result in a lifetime marriage +1

**Comment by assiras** [comment]
👍🏻  +1 Maye, this could be the right time to get deeper into Java and send a PR
